---
title: "Analysis Correlations of features associated with genes DE and detailed PCA 125 patients"
author: "Carla Casanova"
output: pdf_document
date: '2022-06-14'
header-includes:
  - |
    ```{=latex}
    \usepackage{fvextra}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{
      breaksymbolleft={}, 
      showspaces = false,
      showtabs = false,
      breaklines,
      commandchars=\\\{\}
    }
    ```
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 80), tidy = TRUE)
```


```{r libraries}
library(factoextra)
library(corrplot)
library(FactoMineR)
library(NMF)
library(RColorBrewer)
library(NbClust)
```

```{r load_data}
load("/Users/carlacasanovasuarez/Documents/Master Bioinformatics UAB/Prácticas Radiomics/Radiomic features/Results_rfeatures/R objects/rdr_assay_scaled.rda")
load("/Users/carlacasanovasuarez/Documents/Master Bioinformatics UAB/Prácticas Radiomics/Radiomic features/Results_rfeatures/R objects/sputum_eset_countsOK.rda")
load("/Users/carlacasanovasuarez/Documents/Master Bioinformatics UAB/Prácticas Radiomics/Radiomic features/Results_rfeatures/R objects/sputum_eset_phenoOK.rda")
```


```{r check_assay}
head(rdr_assay, 3)
```


## Corralation analysis: features with positive DE  

```{r positive_features_Combat}
pos_rf <- list("MeshVolume.original",
               "Sphericity.original",
               "VoxelVolume.original",
               "Variance.original",
               "Autocorrelation.original",
               "ClusterShade.original",
               "JointAverage.original",
               "SumAverage.original",
               "HighGrayLevelEmphasis.original",
               "LongRunLowGrayLevelEmphasis.original",
               "HighGrayLevelZoneEmphasis.original",
               "LargeAreaEmphasis.original",
               "LowGrayLevelEmphasis.original",
               "LargeAreaHighGrayLevelEmphasis.original",
               "LargeAreaLowGrayLevelEmphasis.original",
               "LowGrayLevelZoneEmphasis.original",
               "SmallAreaEmphasis.original",
               "SmallAreaHighGrayLevelEmphasis.original",
               "SmallAreaLowGrayLevelEmphasis.original",
               "ZoneVariance.original",
               "DependenceNonUniformity.original",
               "LargeDependenceHighGrayLevelEmphasis.original",
               "LargeDependenceLowGrayLevelEmphasis.original",
               "Busyness.original",
               "Coarseness.original",
               "Strength.original",
               "InterquartileRange.original",
               "Range.original")
```

Subset original data of patients with transcriptomics:  

```{r subset}
# Prepare data in object: rdr_assay.scaled
rownames(rdr_assay)[rownames(rdr_assay) == "10Percentile.original"] <- "X10Percentile.original"
rownames(rdr_assay)[rownames(rdr_assay) == "90Percentile.original"] <- "X90Percentile.original"

# Store an array with values of features associated to genes DE
rdr_positive <- rdr_assay[unlist(pos_rf),]

# Change labels to visualize better
rownames(rdr_positive) <- substr(rownames(rdr_positive), start = 1, stop =  nchar(rownames(rdr_positive))-9)
```

See correlations between features positively associated to genes DE:  

```{r corr_matrix}
cor.mat <- round(cor(t(rdr_positive)),2)

head(cor.mat,4)
```


```{r corrplot}
#pdf("Radiomic features from DE analysis correlation matrix (ComBat good).pdf")

corrplot(cor.mat, type="upper", order="hclust", 
         tl.col="black", tl.srt=45, tl.cex = 0.4)

#dev.off()
```

## HCL  

Variables must be columns and observations row:  

```{r heatmap}
#pdf("Heatmap positively correlated features with genes DE.pdf")

# Colors
mypalette <- brewer.pal(11, "PiYG")
morecols <- colorRampPalette(mypalette)

# Heatmap
aheatmap(rdr_positive, col=rev(morecols(37)), main = "Rfeatures with genes DE", scale = "none")

#dev.off()
```


```{r elbow_silhouette}
set.seed(123)

#pdf("Silhouette method optimal k.pdf")

fviz_nbclust(rdr_positive, kmeans, method = "silhouette", k.max = 20) + 
  labs(subtitle = "Silhouette method")

#dev.off()
```


```{r dendo_rf_kmeans}
#pdf("Dendogram positive features clustering (ComBat) k6 kmeans.pdf")

# Compute hierarchical clustering and cut into 4 clusters
#res <- hcut(rdr_positive, k = 6, stand = TRUE)
res <- hkmeans(rdr_positive, k = 6)
# Visualize
# k_colors = c("#00AFBB","#2E9FDF", "#E7B800", "#FC4E07")
fviz_dend(res, rect = TRUE, cex = 0.4, lwd = 0.4, labels_track_height = 20,
          main = "Radiomic features associated with DE genes (sputum samples)", palette = "lancet")

#dev.off()
```


```{r clusters_rf_kmeans_plot}
k.res <- kmeans(rdr_positive, centers = 6)

#pdf("Clustering positive features (ComBat) k6.pdf")
fviz_cluster(k.res, 
             geom = "text", data = rdr_positive,
             ggtheme = theme_minimal(),
             main = "Clustering radiomic features associated with DE genes from sputum",
             repel = TRUE) 
#dev.off()
```







